src/libraries/System.Private.Xml/src/System/Xml/Xsl/XPathConvert.cs

```diff
@@ -2961,236 +2961,14 @@ public static string DoubleToString(double dbl)
             }
         }
 
-        private static bool IsAsciiDigit(char ch)
-        {
-            return unchecked((uint)(ch - '0')) <= 9;
-        }
-
-        private static bool IsWhitespace(char ch)
-        {
-            return ch == '\x20' || ch == '\x9' || ch == '\xA' || ch == '\xD';
-        }
-
-        private static unsafe char* SkipWhitespace(char* pch)
-        {
-            while (IsWhitespace(*pch))
-            {
-                pch++;
-            }
-            return pch;
-        }
-
-        /*  ----------------------------------------------------------------------------
-            StringToDouble()
-
-            Converts a string to a floating point number according to XPath rules.
-            NaN is returned if the entire string is not a valid number.
-
-            This code was stolen from MSXML6 DecimalFormat::parse(). The implementation
-            depends on the fact that the String objects are always zero-terminated.
-        */
         public static unsafe double StringToDouble(string s)
         {
-            Debug.Assert(('0' & 0xF) == 0, "We use (ch & 0xF) to convert char to digit");
-            // For the mantissa digits. After leaving the state machine, pchFirstDig
-            // points to the first digit and pch points just past the last digit.
-            // numDig is the number of digits. pch - pchFirstDig may be numDig + 1
-            // (if there is a decimal point).
-
-            fixed (char* pchStart = s)
+            if (double.TryParse(s, NumberStyles.Float, CultureInfo.InvariantCulture, out double result) && double.IsFinite(result))
             {
-                int numDig = 0;
-                char* pch = pchStart;
-                char* pchFirstDig = null;
-                char ch;
-
-                int sign = 1;       // sign of the mantissa
-                int expAdj = 0;     // exponent adjustment
-
-            // Enter the state machine
-
-            // Initialization
-            LRestart:
-                ch = *pch++;
-                if (IsAsciiDigit(ch))
-                {
-                    goto LGetLeft;
-                }
-
-                switch (ch)
-                {
-                    case '-':
-                        if (sign < 0)
-                        {
-                            break;
-                        }
-                        sign = -1;
-                        goto LRestart;
-                    case '.':
-                        if (IsAsciiDigit(*pch))
-                        {
-                            goto LGetRight;
-                        }
-                        break;
-                    default:
-                        // MSXML has a bug, we should not allow whitespace after a minus sign
-                        if (IsWhitespace(ch) && sign > 0)
-                        {
-                            pch = SkipWhitespace(pch);
-                            goto LRestart;
-                        }
```

Comment:
@krwq This is the bug. If there is whitespace after minus sign, the state of parser is reset in MSXML.

Reply:
I think this change is ok, any other diffs you're aware of? I'd prefer we've added some tests here